classes/local/store/object_file_system.php

```diff
@@ -1154,4 +1163,51 @@ private function update_object(array $result): array {
 
         return $result;
     }
+
+    /**
+     * Pushes tags to the external store (post upload) for a given hash.
+     * External client must support tagging.
+     *
+     * @param string $contenthash file to sync tags for
+     */
+    public function push_object_tags(string $contenthash) {
+        if (!$this->get_external_client()->supports_object_tagging()) {
+            throw new coding_exception("Cannot sync tags, external client does not support tagging.");
+        }
+
+        // Get a lock before syncing, to ensure other parts of objectfs are not moving/interacting with this object.
+        $lock = $this->acquire_object_lock($contenthash, 10);
+
+        // No lock - just skip it.
+        if (!$lock) {
+            throw new coding_exception("Could not get object lock");
+        }
```

Comment:
I'd like to understand the failure mode a bit more here. Ideally if we can't get a lock then something else is doing stuff, and it might be doing the same stuff. So we could wait with a 0 timeout, and then just keep moving. The question is will this then get picked up on the next cron run and retried?

Reply:
It's an edge case. If this is hit by the migration task (where I think it will be hit the most) it will retry since the sync status will still be 'needs sync', but if this is hit in the course of normal file upload I think it will just stay that way until the migration task runs.

I imagine a scenario you could hit this is if you are uploading a really big file, and the record exists in mdl_files but objectfs is still uploading it, and the tagging migration task tries to tag it.

Reply:
If you can't get a lock this should be business as usual, this should not be a coding_exception which by definition implies the code needs to be fixed

I don't think this should be an exception at all as it would break the file interaction, we want to just ignore it and move on and assume that it will get picked up again later by cron?